posts/MarkoB/2018-01-04-adding-new-constraint-to-engine.adoc

```diff
@@ -0,0 +1,488 @@
+
+== Step 1 - Add constraint annotation and validator
+
+**Commit:** https://github.com/hibernate/hibernate-validator/commit/5ba6668ef0725db718d98a9f34cb9cd4bcaa4a3e
+
+The first main step is to create the constraint itself with corresponding validators and
+register them in the engine.
+
+=== Add constraint annotation
+
+In order to do this, first a new constraint annotation should be created. In our case it will be called
+`@ISBN` and have one attribute of its own, `type`, which would define if the ISBN under inspection is
+ISBN 10 or ISBN 13.
+
+```java
+/**
+ * Checks that the annotated character sequence is a valid
+ * <a href="https://en.wikipedia.org/wiki/International_Standard_Book_Number">ISBN</a>.
+ * The length of the number and the check digit are both verified.
+ * <p>
+ * The supported type is {@code CharSequence}. {@code null} is considered valid.
+ * <p>
+ * During validation all non ISBN characters are ignored. All digits and 'X' are considered
+ * to be valid ISBN characters. This is useful when validating ISBN with dashes separating
+ * parts of the number (ex. {@code 978-161-729-045-9}).
+ *
+ * @author Marko Bekhta
+ * @since 6.0.6
+ */
+@Documented
+@Constraint(validatedBy = { })
+@Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER, TYPE_USE })
+@Retention(RUNTIME)
+@Repeatable(List.class)
+public @interface ISBN {
+    String message() default "{org.hibernate.validator.constraints.ISBN.message}";
+
+    Class<?>[] groups() default { };
+
+    Class<? extends Payload>[] payload() default { };
+
+    Type type() default Type.ISBN13;
+
+    /**
+     * Defines several {@code @ISBN} annotations on the same element.
+     */
+    @Target({ METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER, TYPE_USE })
+    @Retention(RUNTIME)
+    @Documented
+    public @interface List {
+        ISBN[] value();
+    }
+
+    /**
+     * Defines the ISBN length. Valid lengths of ISBNs are {@code 10} and {@code 13}
+     * which are represented as {@link Type#ISBN10} and {@link Type#ISBN13} correspondingly.
+     */
+    enum Type {
+        ISBN10,
+        ISBN13
+    }
+}
+```
+
+Looking at this code snippet, it is worth emphasizing the following:
+
+- An empty array as `validatedBy` parameter in `@Constraint`. There is no need to declare
+any validators here as they will be registered in the engine directly.
+- Key for default message is in the format `{org.hibernate.validator.constraints.${nameOfConstraint}.message}`,
+where `${nameOfConstraint}` should be replaced with the same string as the name of constraint.
+- If applicable, a `@Repeatable` annotation should be defined.
+- New constraints are added to `org.hibernate.validator.constraints` package or its subpackages.
+- It is important to add a javadoc describing what kind of checks are performed by the constraint
+as well as to which types it is applicable.
+- An `@since` tag should specify the version in which we introduce the constraint.
```

Comment:
I think it should be s/An/A

Reply:
Yeah, in my head, it sounds like "at since", that's why I use "an". But I think you're right, fixed.